clients/zaway/zaway.1
.\"	$Id: zaway.1,v 1.9 1999-07-21 12:34:18 ghudson Exp $
.\"
.\" Copyright 1987,1988 by the Massachusetts Institute of Technology
.\" All rights reserved.  The file /usr/include/zephyr/mit-copyright.h
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)zaway.1	6.1 (MIT) 7/9/87
.\"
.TH ZAWAY 1 "July 1, 1988" "MIT Project Athena"
.ds ]W MIT Project Athena
.SH NAME
zaway \- tell other people via Zephyr that you aren't around
.SH SYNOPSIS
.B zaway
[
.I OPTIONS
]
[
.I FILE
]
.SH DESCRIPTION
.I zaway
provides a way for you to automatically send replies when other people
contact you using
.I zwrite(1). zaway
subscribes itself to class "MESSAGE", instance "*", so that it can
monitor your incoming messages.  It does not affect the operation of any
other client receiving messages.
.I zaway
is typically run when you are leaving your terminal or display
temporarily.
.I zaway
usually never exits; when you return to your terminal you should type
the interrupt character (usually ^C) in order to make
.I zaway
exit.
.SS OPTIONS
.TP
.I "\-m STRING"
Use STRING as the body of the auto-reply message.  Any message file
(specified on the command line or the default) is ignored.
.TP
.I "\-w"
Watch the invoking user's location status.  If the user is locatable
anywhere, no auto-replies will be sent.
.TP
.I "\-h"
Displays a short usage message and exits.
.PP
.I zaway
uses a message file (which defaults
to $HOME/.away) to describe what reponses should be sent to which
senders.  The general format of this file
is:
.PP
.nf
	>name
	>name
	message
	>name
	message
.fi
.PP
Any number of user names may be specified preceding the message to send
to those senders.  If a user name appears more than once, the message will
be a concatenation of each of the appropriate messages.  There are
two special names: "*" indicates that the following message should be
sent to all senders and "%" indicates that the following message should
only be sent if the user name has not matched yet. 
.PP
If no file is specified,
and no default file can be found, the following message is returned:
.sp
.in +5
I'm sorry, but I am currently away from the terminal and am
not able to receive your message.
.in -5
.sp
If a user name does not match any of those listed in the file, and no
"*" or "%" field is specified, no return message is sent.
All messages are preceded by the line "Automated reply:".
To avoid loops, messages are not sent in response to messages beginning
with "Automated reply:" or sent by the same Kerberos principal as the
user running 
.I zaway.

.SH SAMPLE FILE
.nf
>eichin
>tony
Hi there guys!  I'm in the other room right now.
I'll be back in 5 minutes or so.
>jruser
Sorry, but I'm gone for the day...
>%
Hello...I'm not sure who you are.  I'll be back soon,
though.
>*
This message comes to you compliments of zaway!
.fi

The final "compliments" message will be included in all messages,
whereas the "I'm not sure" message will only be included in messages that
are not from "eichin", "tony", or "jruser".
.SH FILES
$HOME/.away
.SH SEE ALSO
zephyr(1), zwgc(1), zwrite(1), zhm(8), zephyrd(8)
.br
Project Athena Technical Plan Section E.4.1, `Zephyr Notification
Service'
.SH AUTHOR
.PP
Robert S. French (MIT-Project Athena)
.SH RESTRICTIONS
Copyright (c) 1987,1988 by the Massachusetts Institute of Technology.
All Rights Reserved.
.br
.I zephyr(1)
specifies the terms and conditions for redistribution.


clients/zaway/zaway.c
/* This file is part of the Project Athena Zephyr Notification System.
 * It contains code for the "zaway" command.
 *
 *	Created by:	Robert French
 *
 *	$Source: /afs/dev.mit.edu/source/repository/athena/lib/zephyr/clients/zaway/zaway.c,v $
 *	$Author: rfrench $
 *
 *	Copyright (c) 1987 by the Massachusetts Institute of Technology.
 *	For copying and distribution information, see the file
 *	"mit-copyright.h". 
 */

#include <zephyr/mit-copyright.h>

#include <zephyr/zephyr.h>

#include <pwd.h>

#ifndef lint
static char rcsid_zaway_c[] = "$Header: /afs/dev.mit.edu/source/repository/athena/lib/zephyr/clients/zaway/zaway.c,v 1.1 1987-08-06 22:34:37 rfrench Exp $";
#endif lint

#define MESSAGE_CLASS "MESSAGE"
#define DEFAULT_MSG "I'm sorry, but I am currently away from the terminal and am\nnot able to receive your message.\n"

main(argc,argv)
	int argc;
	char *argv[];
{
	FILE *fp;
	ZNotice_t notice;
	ZPacket_t packet;
	ZSubscription_t sub;
	u_short port;
	int retval;
	struct passwd *pw;
	char awayfile[BUFSIZ],inrep[BUFSIZ],*ptr,*msg[2],*envptr;
	char *find_message();
	
	if ((retval = ZInitialize()) != ZERR_NONE) {
		com_err(argv[0],retval,"while initializing");
		exit(1);
	}

	port = 0;
	if ((retval = ZOpenPort(&port)) != ZERR_NONE) {
		com_err(argv[0],retval,"while opening port");
		exit(1);
	}

	sub.class = MESSAGE_CLASS;
	sub.classinst = "*";
	sub.recipient = ZGetSender();

	if (argc > 1)
		strcpy(awayfile,argv[1]);
	else {
		envptr = (char *)getenv("HOME");
		if (envptr)
			sprintf(awayfile,"%s/.away",envptr);
		else {
			if (!(pw = getpwuid(getuid()))) {
				fprintf(stderr,"Who are you?\n");
				exit(1);
			}
			sprintf(awayfile,"%s/.away",pw->pw_dir);
		} 
	}

	fp = fopen(awayfile,"r");
	if (!fp && argc > 1) {
		fprintf(stderr,"File %s not found!\n",awayfile);
		exit(1);
	} 
	
	if ((retval = ZSubscribeTo(&sub,1,port)) != ZERR_NONE) {
		com_err(argv[0],retval,"while subscribing");
		exit(1);
	}

	for (;;) {
		if ((retval = ZReceiveNotice(packet,sizeof packet,
					     &notice,0)) != ZERR_NONE) {
			com_err(argv[0],retval,"while receiving notice");
			continue;
		}
		if (!strcmp(notice.z_sender,ZGetSender()))
			continue;
		if (!strcmp(notice.z_message,"Automated reply:\n"))
			continue;
		if (fp) {
			if (!(ptr = find_message(&notice,fp)))
				continue;
		}
		else {
			ptr = (char *)malloc(sizeof(DEFAULT_MSG)+1);
			if (!ptr) {
				com_err(argv[0],errno,"while getting default message");
				exit(1);
			}
			strcpy(ptr,DEFAULT_MSG);
		}
		notice.z_recipient = notice.z_sender;
		notice.z_sender = 0;
		notice.z_default_format = 0;
		     
		msg[0] = "Automated reply:\n";
		msg[1] = ptr;
		
		notice.z_message_len = strlen(notice.z_message)+1;
		if ((retval = ZSendList(&notice,msg,2,ZNOAUTH)) != ZERR_NONE) {
			com_err(argv[0],retval,"while sending notice");
			continue;
		}
		free(ptr);
	}
}

char *find_message(notice,fp)
	ZNotice_t *notice;
	FILE *fp;
{
	char bfr[BUFSIZ],*ptr,*ptr2,sender[BUFSIZ];
	int gotone,lastwasnt;
	char *realloc();
	
	rewind(fp);

	strcpy(sender,notice->z_sender);
	ptr2 = (char *)index(sender,'@');
	if (ptr2)
		*ptr2 = '\0';
	
	ptr = 0;
	gotone = 0;
	lastwasnt = 0;
	
	while (fgets(bfr,sizeof bfr,fp) > 0) {
		if (*bfr == '>') {
			if (lastwasnt)
				gotone = 0;
			bfr[strlen(bfr)-1] = '\0';
			ptr2 = (char *)index(bfr,'@');
			if (ptr2)
				*ptr2 = '\0';
			if (!strcmp(bfr+1,sender) ||
			    !strcmp(bfr+1,"*") ||
			    (!strcmp(bfr+1,"%") && !ptr))
				gotone = 1;
			lastwasnt = 0;
		} 
		else {
			if (gotone) {
				if (!ptr) {
					ptr = (char *)malloc(strlen(bfr)+1);
					*ptr = '\0';
				} 
				else
					ptr = (char *)realloc(ptr,strlen(bfr)+
							      strlen(ptr)+1);
				strcat(ptr,bfr);
			}
			lastwasnt = 1;
		}
	}

	return (ptr);
}
